OpenKeychain/src/main/java/org/sufficientlysecure/keychain/ui/KeyListFragment.java

```diff
@@ -292,25 +318,73 @@ private void startSearchForQuery() {
         // This is called when a new Loader needs to be created. This
         // sample only has one Loader, so we don't care about the ID.
         Uri uri;
-        if (!TextUtils.isEmpty(mQuery)) {
-            uri = KeyRings.buildUnifiedKeyRingsFindByUserIdUri(mQuery);
+        if (!TextUtils.isEmpty(getQuery())) {
+            uri = KeyRings.buildUnifiedKeyRingsFindByUserIdUri(getQuery());
         } else {
             uri = KeyRings.buildUnifiedKeyRingsUri();
         }
 
         // Now create and return a CursorLoader that will take care of
         // creating a Cursor for the data being displayed.
         return new CursorLoader(getActivity(), uri,
-                KeySectionedListAdapter.KeyListCursor.PROJECTION, null, null,
-                KeySectionedListAdapter.KeyListCursor.ORDER);
+                PROJECTION, null, null,
+                ORDER);
     }
 
+    public static final String ORDER = KeychainContract.KeyRings.HAS_ANY_SECRET
+            + " DESC, " + KeychainContract.KeyRings.USER_ID + " COLLATE NOCASE ASC";
+
+    public static final String[] PROJECTION = {
+            KeyRings._ID,
+            KeychainContract.KeyRings.MASTER_KEY_ID,
+            KeychainContract.KeyRings.USER_ID,
+            KeychainContract.KeyRings.IS_REVOKED,
+            KeychainContract.KeyRings.IS_EXPIRED,
+            KeychainContract.KeyRings.IS_SECURE,
+            KeychainContract.KeyRings.HAS_DUPLICATE_USER_ID,
+            KeychainContract.KeyRings.CREATION,
+            KeychainContract.KeyRings.NAME,
+            KeychainContract.KeyRings.EMAIL,
+            KeychainContract.KeyRings.COMMENT,
+            KeychainContract.KeyRings.VERIFIED,
+            KeychainContract.KeyRings.HAS_ANY_SECRET,
+            KeychainContract.KeyRings.FINGERPRINT,
+            KeychainContract.KeyRings.HAS_ENCRYPT};
+
+
     @Override
     public void onLoadFinished(Loader<Cursor> loader, Cursor data) {
-        // Swap the new cursor in. (The framework will take care of closing the
-        // old cursor once we return.)
-        getAdapter().setSearchQuery(mQuery);
-        getAdapter().swapCursor(KeySectionedListAdapter.KeyListCursor.wrap(data));
+
+        mKeyItems.clear();
+
+        if (data.moveToFirst()) {
+            while (!data.isAfterLast()) {
+                mKeyItems.add(new KeyItem(null, KeyListCursor.wrap(data), mKeyListener));
+                data.moveToNext();
+            }
+        }
```

Comment:
This is not the way we want to handle this. This code loads all items into an ArrayList before using the ArrayList from memory to display the items. For a long list of keys this requires that all keys are loaded into the memory, which is not good. Instead, the adapter should be able to work on a Cursor directly (as done before).

Unfortunately, this is a major blocker which needs to be resolved first.